rllib/env/remote_base_env.py

```diff
@@ -49,71 +55,88 @@ def __init__(self,
             existing_envs: Optional list of already created sub-environments.
                 These will be used as-is and only as many new sub-envs as
                 necessary (`num_envs - len(existing_envs)`) will be created.
+            worker: An optional RolloutWorker that owns the env. This is only
+                used if `remote_worker_envs` is True in your config and the
+                `on_sub_environment_created` custom callback needs to be
+                called on each created actor.
         """
 
         # Could be creating local or remote envs.
         self.make_env = make_env
-        # Whether the given `make_env` callable already returns ray.remote
-        # objects or not.
-        self.make_env_creates_actors = False
-        # Already existing env objects (generated by the RolloutWorker).
-        self.existing_envs = existing_envs or []
         self.num_envs = num_envs
         self.multiagent = multiagent
         self.poll_timeout = remote_env_batch_wait_ms / 1000
+        # Already existing env objects (generated by the RolloutWorker).
+        self.existing_envs = existing_envs or []
+        self.worker = worker
+
+        # Whether the given `make_env` callable already returns ray.remote
+        # objects or not.
+        self.make_env_creates_actors = False
+
+        self._observation_space = None
+        self._action_space = None
 
         # List of ray actor handles (each handle points to one @ray.remote
         # sub-environment).
         self.actors: Optional[List[ray.actor.ActorHandle]] = None
-        self._observation_space = None
-        self._action_space = None
+
+        # `self.make_env` already produces Actors: Use it directly.
+        if len(self.existing_envs) > 0 and isinstance(self.existing_envs[0],
+                                                      ray.actor.ActorHandle):
+            self.make_env_creates_actors = True
+            self.actors = []
+            while len(self.actors) < self.num_envs:
+                sub_env = self.make_env(len(self.actors))
+                if self.worker is not None:
+                    self.worker.callbacks.on_sub_environment_created(
+                        worker=self.worker,
+                        sub_environment=sub_env,
+                        env_context=self.worker.env_context.
+                        copy_with_overrides(vector_index=len(self.actors)))
+                self.actors.append(sub_env)
+        # `self.make_env` produces gym.Envs (or children thereof, such
+        # as MultiAgentEnv): Need to auto-wrap it here. The problem with
+        # this is that custom methods wil get lost. If you would like to
+        # keep your custom methods in your envs, you should provide the
+        # env class directly in your config (w/o tune.register_env()),
+        # such that your class will directly be made a @ray.remote
+        # (w/o the wrapping via `_Remote[Multi|Single]AgentEnv`).
+        else:
```

Comment:
wait, we can also get here if len(self.existing_envs) == 0.
is that not a useful condition? should we get rid of it from if above?

Reply:
Fixed now.

I hope the docstring is clear enough now on the `existing_envs` arg:

```
            existing_envs: Optional list of already created sub-environments.
                These will be used as-is and only as many new sub-envs as
                necessary (`num_envs - len(existing_envs)`) will be created.
```

This is all due to the fact that a RolloutWorker first creates a single sub-env up front, just to see what this is (already a VectorEnv, a gym.Env, a ActorHandle (remote envs), etc..) to then decide what to do next (plus some validation).

Reply:
This sub-env then must be re-cycled when we do the vectorization, e.g. inside `RemoteBaseEnv` or other vectorized RLlib env classes.

Reply:
get it get it.
I don't remember where we recycle this sub-env during vectorization actually. hopefully it's done properly and enforced :)